Optionally use multiprocessing.Queue for MWorker to publisher

dwoz opened the pull request:
### What does this PR do?

Adds an option to use a `multiprocessing.Queue` instead of ZeroMQ PUSH/PULL sockets for messages relayed from MWorkers to the publisher daemon process.

### What issues does this PR fix or reference?

#36469
#46553
#50393

### Previous Behavior

ZeroMQ's PUSH/PULL sockets are unreliable when dealing with large messages (size) or large number of messages sent quickly.

### New Behavior

Optionally a multiprocessing.Queue for messages moving from MWorkers to the publisher daemon. Messages will get published reliably.

### Tests written?

No - This patch has been tested on via jenkin's branch tests on a fork of 2017.7

### Commits signed with GPG?

Yes

dwoz committed:
Use multiprocessing queue for pub server msgs

dwoz committed:
Add multiprocessing queue config for pubserver

dwoz committed:
Always configure pub_sock

dwoz committed:
Fix config option type

dwoz committed:
Add documentation for new options

dwoz committed:
Use mp queue in test suite

dwoz committed:
Fix linter errors/warnings

rares-pop commented on salt/transport/zeromq.py:
```diff
@@ -905,12 +908,14 @@ def publish(self, load):
+            msg_queue.put(payload)
+        else:
+            pub_sock.send(payload)
+            pub_sock.close()
```
should we cache the socket and not create it and close it for each message we publish?

dwoz replied:
I did try storing the context and socket on the class so it won't get created each time but I still saw the issue. @kingsquirrel152 also talks about trying that as a fix in #36469

kingsquirrel152 replied:
I fixed this issue in an older version of salt stack (back when it was single threaded and not driven by tornado). The issue iirc is that creating the socket, and sending data, then _tearing the socket down_, breaks the contract with zmq. You have no guarantee the message is ever sent. I think there are some settings for the socket that could help (linger for instance). TLDR though - I think your best bet is to follow @rares-pop suggest and cache the socket avoiding the issue entirely.

cachedout commented:
@DmitryKuzmenko What do you think of this?

DmitryKuzmenko commented:
@cachedout I already had a quick lookup of this PR today. Just thought that there is some reason to do this that I don't know. But if there isn't here is my opinion.

ZMQ and Tornado TCP/IPC are always faster than MP Queues. So we need a reasonable argument to use queues instead of ipc/tcp. I'll take a closer look on that related issues tomorrow.

dwoz commented:
@DmitryKuzmenko I added a muliprocessing.Queue here because I was un-able to make the zeromq PUSH/PULL reliable. In the windows test suite we are seeing tests fail after messages get dropped while moving over the PUSH/PULL [from MWorkers](https://github.com/saltstack/salt/blob/2017.7/salt/transport/zeromq.py#L908) to the [publish daemon](https://github.com/saltstack/salt/blob/2017.7/salt/transport/zeromq.py#L805). I recently added the debug level logs around those calls to more easily spot this when it happens in the test suites.

I tried adjusting the zeromq watermarks. I also moved context and socket to the class level so they only created once, something @rares-pop mentions in his comment. I still saw the dropped messages with these changes. 

After, I moved to a multiprocessing.Queue. I agree that sticking with zmq will be more performant and that it the primary reason I made the Queue optional. I also think it would still be nice if we can make zmq rock solid. I'm re-testing the class level context just to be on the safe side that I didn't miss anything before. The only other thing I can think of is to replace PUSH/PULL with PUB/SUB in hopes of that being more reliable. It seems like that would still work if we only have one subscriber (subscribed to everything) with multiple publisher.

I am looking forward to hearing more of your thoughts on this if you have them. Thanks for looking.

dwoz commented:
I've opened a new PR (#50463) which I think resolves these issues without using `multiprocessing.Queue` and has some tests. I'm going to close this one since the other seems to be reliable. Thanks for everyone's input.
